wpilib/wpilib/watchdog.py

```diff
@@ -75,18 +171,73 @@ def reset(self) -> None:
 
     def enable(self) -> None:
         """Enables the watchdog timer."""
-        self.startTime = Timer.getFPGATimestamp()
+        self._startTime = hal.getFPGATime()
         self._isExpired = False
-        self.epochs.clear()
-        self.notifier.startPeriodic(self.timeout)
+        self._epochs.clear()
+
+        with self._queueMutex:
+            self._isExpired = False
+
+            watchdogs = self._watchdogs
+            try:
+                watchdogs.remove(self)
+            except ValueError:
+                pass
+            else:
+                heapq.heapify(watchdogs)
+
+            self._expirationTime = self._startTime + self._timeout
+            heapq.heappush(watchdogs, self)
+            self._schedulerWaiter.notify_all()
 
     def disable(self) -> None:
-        """Disable the watchdog."""
-        self.notifier.stop()
-
-    def timeoutFunc(self) -> None:
-        if not self._isExpired:
-            logger.info("Watchdog not fed after %ss" % (self.timeout,))
-            self.callback()
-            self._isExpired = True
-            self.disable()
+        """Disables the watchdog timer."""
+        with self._queueMutex:
+            self._isExpired = False
+
+            watchdogs = self._watchdogs
+            try:
+                watchdogs.remove(self)
+            except ValueError:
+                pass
+            else:
+                heapq.heapify(watchdogs)
+            self._schedulerWaiter.notify_all()
+
+    @classmethod
+    def _schedulerFunc(cls) -> None:
+        getFPGATime = hal.getFPGATime
+
+        with cls._queueMutex:
+            while cls._keepAlive:
+                if cls._watchdogs:
+                    delta = cls._watchdogs[0]._expirationTime - getFPGATime()
+                    timedOut = cls._schedulerWaiter.wait(delta)
```

Comment:
Condition variables return False when timed out, maybe that's why you weren't able to get the tests to work?